GrainTracker handle no grain (void region) or separate phase case

permcody opened the pull request:
The GrainTracker change was really simple. This gets us a little closer.

@dschwen - This example is a little problematic because the GrainTracker is printing "re-assigning grain"! It's time to "fix" how the grain tracker assigns the EBSD information, but I'm not sure how to do that reliably. I'm coming over to chat some more, prepare yourself.

Commit:
Fixed EulerAngleProvider2RGB to work with areas not assigned to a grain for #7384

Commit:
Added test with non-grain regions for #7384

permcody committed:
Adding phase option to GrainTracker

closes #7384

permcody committed:
Re-write EBSD logic completely

permcody commented:
Yikes, another huge PR... I'll do more extensive testing this weekend and early next week. This should be a lot more reliable with the EBSD cases and hopefully doesn't change anything for the Voronoi cases. @frombs, @tonkmr - feel free to pull this branch before it merges and see if it works for you

frombs commented:
@permcody - I see you made lots of changes to the framework.  Thanks for your efforts to get this PR pushed through.

I pulled the branch and ran my dataset.  The new version creates 697 features that are tracked by GrainTracker.  The previous version created 676 features.  The EBSD dataset contains 670 grains plus the void region (671 total features).  After the first time step, I noticed a lot of grains that were marked inactive and lots of new grains that were created by GrainTracker.   I am attaching my input files so you can run it.  It takes about 3 minutes to run with "-n 12" but you can reduce it by adjusting the number of time steps.  Note: I had to add the "txt" extension to the "i" file so that github would upload it.

Would it be possible to adjust the "feature_volume_file" to add zeros for features that become inactive in this PR?  I am also not sure how to utilize the data in the file.  I need to track the volumes of the features (grains & voids) found in the EBSD datafile.  Should there be a 1 to 1 correlation for features within GrainTracker and the EBSD data features?

@tonkmr - I figured out how to get good convergence using the hypre boomeramg preconditioner for my simulation (split CH).  You can see the settings in the input file.  I am not sure if it works only for my dataset or in general, but it will help when running the larger simulations on Falcon.

Input files:
[UO2-524_208_196_Marmot_Slice3.txt](https://github.com/idaholab/moose/files/440964/UO2-524_208_196_Marmot_Slice3.txt)
[coarsen_only.i.txt](https://github.com/idaholab/moose/files/440966/coarsen_only.i.txt)

permcody commented:
@tonkmr or @frombs - I know you guys are getting pretty good at processing these EBSD files. Can one of you make a small test case that has one of these split EBSD grains (similar to the existing EBSD test). If not, I'll go in and hack the existing case and reuse one of the existing grain numbers. This split grain stuff is still tripping me up. We need a test case so that it doesn't break anymore. It seems that most of your larger scans have these split grains in them so they are very real.

permcody commented:
Oh maybe I don't need that. The new test file that Mike put in has one (I think). It's a little on the large side but I think it'll work

permcody committed:
Cleanup and rename FeatureData ID

frombs commented:
@permcody - Below are links to a new test file that was cropped from the larger dataset.  It is a 40 x 40 grid with 35 grains and contains one split grain that borders the void region.  It only takes a couple minutes to create the file so let me know what grid size you want and I will create it for you.

[splitCH_GrainTracker_test.i.txt](https://github.com/idaholab/moose/files/441699/splitCH_GrainTracker_test.i.txt)
[UO2-524_40_40_Marmot.txt](https://github.com/idaholab/moose/files/441697/UO2-524_40_40_Marmot.txt)

permcody commented:
Thanks Brad! Be careful about putting any kind of UO2 data in publicly accessible locations. I've got this dataset but you might want to pull it back down. Maybe rename it at the very least?

permcody commented:
Ugh, I just realized that Mike's PR (which this one is based on) has a UO2 data set in it as well. I'll take care of that one.

BTW - That data set is pretty good - it's absolutely terrible in that it has several single element grains which was one of the big concerns I had. Those are ridiculously hard to handle since they are basically indistinguishable from adjacent grains that surround the single element grains on two or more sides once we've weighted the contributions on the nodes. Oh well, it needs to work since this is what real data looks like. I've thought about this a lot and I think I have another idea up my sleeve. The only thing I might suggest for helping me is to make the input files as simple as possible. There are a few Marmot objects in this one so I just re-used the input file that Mike posted instead. I don't really care about the physics, I just need it to run.

frombs commented:
@permcody and @dschwen - thanks for your efforts to move this PR forward.

@permcody - Thanks for catching the issue with the UO2 data.  The updated file links can be found below.  I removed all references to UO2 but it still uses Marmot due to the SplitCHPoreGB kernel and PoreGBMtrxMobMaterial material.  I also simplified the input file by removing unnecessary aux variables and postprocessors.  I reduced the time steps from 10 to 3 so it only takes 10 seconds to run on a single processor.  Let me know if you need anything else.

[splitCH_GrainTracker_test.i.txt](https://github.com/idaholab/moose/files/443339/splitCH_GrainTracker_test.i.txt)
[splitCH_40x40_Marmot.txt](https://github.com/idaholab/moose/files/443340/splitCH_40x40_Marmot.txt)

permcody commented:
I just wanted to post an update. This 40x40 data set = is terrible, which is good news. I contains so many different edge cases all in one. There are no less than 3 split grains in this file, plus multiple very small grains (2 that are only 1 element). This file essentially proves that our approach with taking elemental data and then creating a "node weight" map cannot possibly work for this kind of data. I'm still contemplating various ideas to make this all work with reduced op modeling which is no small feat.

I've created the ability to create an elemental aux field that is can be populated by `ReconVarIC` that contains the assigned ops (the one piece of data that cannot come from the `EBSDReader`). Theoretically this information should be enough to come up with the perfect grain map in the first time step but I'm still running into a few issues. The single grain case is especially nasty. For one thing the node weight maps of nearby larger grains are actually greater than the contributions to the nodes by the single element grains which it makes it hard to figure out the active grain at a location. I'm finding that I have to look at the order parameter information even less and rely more on the just the EBSDReader + Assigned OP information.  Still lots of work to do before you can run EBSD simulations reliably.

permcody committed:
Add capability to create elemental aux field

The ReconVarIC class can now populate an
elemental aux variable to store the
assigned OPs

tonkmr commented:
I think it would be OK to ignore single element grains entirely, since they will disappear after the first time step. It would probably work fine to have every order parameter set to zero in those elements, because that will also change in the first time step.

Of course, that opens up the question of how big should a grain be before we start to consider it. But I think we could start with just ignoring single element grains.

frombs commented:
I analyzed the raw data and confirmed that there are 2 split grains (grain # 13 and 21) and 3 single element grains (grain # 5, 28, and 37) in the dataset.  Below is an image of the raw data that contains an inverse pole figure color map of the grains with a grain number overlay.  Note that grain 1 is the void region.

![uo2-524_40_40_slice3_ipf_map_with_ids](https://cloud.githubusercontent.com/assets/8171022/18138385/a4bd0f6e-6f69-11e6-8927-4c1455e6a0e8.png)

In the short term (and interest of time) I agree with @tonkmr about ignoring single elements, but for long term work ignoring them could be detrimental in certain cases, such as when dealing with voids and small precipitates.  These features may need to be conserved depending on the situation.  @permcody - Would performing adaptive mesh refinement prior to running grain tracker resolve this issue since they would no longer be single elements?

I also noticed the inverse pole figure coloring is not correct when viewing the "RGB_" object in Paraview.  The indexes and colors appear shifted for many of the grains when compared to the raw data shown above.

permcody committed:
Overhaul ebsd tests

permcody commented:
This PR should be ready to review and it's another huge rewrite. I want to highlight a few things. Rather than determine where grains are based on the IC set from the  `EBSDReader::elem_to_node_map`, I've modified the `ReconVarICAction` object to create an additional elemental field based that will be populated with data that is not interpolated by the EBSDReader. That aux var is used during the initial tracking step along with other `EBSDPointData` to set an exact grain representation handling all (hopefully) nasty edge cases such as single cell grains or split grains. After the initial step, all subsequent time steps use the existing logic for tracking and remapping. This seems to work fine with all of the existing, new and multiphase ebsd data sets.

First, I need feedback on this "extra aux var" that I'm creating. It's not the cleanest way of doing things and I think we can do better moving forward but this is a decent intermediate step to get us to where we need to be for more sane EBSD input files. If that's ok, I think most of the other changes in this PR will be fine.

@frombs - Feel free to grab this branch one more and try it out. 
@dschwen - please review

permcody commented:
Damn - this still isn't ready to merge. I got a seg fault on a larger problem after several time steps. I'll have to track that down before this can be merged.

permcody committed:
Cleanup

permcody committed:
Fixing advanced_op_assignment in EBSD

permcody committed:
INACTIVE grains have been removed from structure

permcody committed:
Avoid unecessary sorting, fixing up APIs

frombs commented:
@permcody - I just tried your newest commit and the simulation completes without the segmentation fault.  It is a lot faster too.  My test file used to take 10 seconds to run and now it runs in only 2.25 seconds.  Nice work!  Thanks for adding the zeros into the Grain Volume file.  Also, It appears that GrainTracker is picking up the correct number of grains.  I will do some more testing in the morning.  

I am still seeing a discrepancy between the raw data and the RGB_ output.  The colors appear to be correct but they are not attached to the correct grain.  See image links below.  If you start at the bottom left and work to the right, the first row of grains have the correct coloring.  The last grain in that row is the void phase (colored black).  The colors in the next row up are shifted starting with the grain shaped like a cross (grain number 14 from the raw dataset).  That particular grain is colored red but there is no red colored grains in the raw data.  Any ideas where that color is coming from?

![uo2-524_40_40_slice3_ipf_map_with_ids](https://cloud.githubusercontent.com/assets/8171022/18138385/a4bd0f6e-6f69-11e6-8927-4c1455e6a0e8.png)

![output](https://cloud.githubusercontent.com/assets/8171022/18238088/51cc5e32-72f6-11e6-8a51-8336c686b4e9.png)

permcody commented:
I'm not sure if I have the exact input file you ran but I just used @tonkmr's test that is added in this PR (euler2rgb_no_grain.i) and loaded this mesh file. I don't see any red grains in the rgb output if I run that input file.

dschwen commented:
Ok, yeah, this indeed looks like somebody somewhere confused _global index_ and _feature id_.

permcody commented:
@dschwen - Will you review this? I'm satisfied with the output on these tests. The marmot diffs are a little more confusing to me. I'm not sure why this PR would trigger the diffs on those tests. They seem small but they really shouldn't be there.

permcody committed:
Avoid logic bugs due to interesting choices of "execute_on"

permcody commented:
Marmot tests should pass with this latest commit.

frombs commented:
@permcody - I clobbered and cleaned my branch and recompiled with your latest commits and then ran Mike's input file.  I am still seeing the red grain (see image below).  I am also seeing variations in the grain shapes between the raw data, my input file, and Mike's input file (see grains # 21, 23, 24, 35, and 37).  These are all along the interface with the void region.  The biggest difference between my input file and Mike's is that he is using the "GBEvolution" material and I am using "PoreGBMtrxMobMaterial."  

![comparison](https://cloud.githubusercontent.com/assets/8171022/18289336/3989c2dc-743c-11e6-8ed2-55c8e156658c.png)

Maybe there is something messed up with my branch if Cody is getting a different result.  Can @tonkmr or @dschwen run these two files and chime in whether they get the same results?

Here is a link to the input files:
[InputFiles.zip](https://github.com/idaholab/moose/files/457659/InputFiles.zip)

permcody commented:
OK - I'm not speaking your language :) 
I guess I don't know what field you're viewing. I thought that you were viewing the RGB field but perhaps you aren't?

If there's a difference it's probably because someone is using the wrong ID somewhere. The number that come back from the grain_tracker in the `UNIQUE_REGIONS` enum is not the feature ID. It's the "global or local id" depending on if you are using the "phase" parameter or not. To get the actual EBSD number you have to use the EBSD reader to report to get that number. If I know what object you are using, we can look at it and see if it's one of these types of mixups.

frombs commented:
I am viewing the "RGB_" field in Paraview with the "Map Scalers" option unchecked.

dschwen commented on modules/phase_field/examples/ebsd_reconstruction/IN100-111grn.i:
```diff
@@ -144,11 +144,11 @@
-    execute_on = 'initial timestep_begin'
+#    threshold = 0.2
+#    connecting_threshold = 0.2
+#    execute_on = 'initial timestep_begin'
```
why not just remove?

dschwen commented on modules/phase_field/include/postprocessors/FeatureFloodCount.h:
```diff
@@ -306,6 +328,18 @@ class FeatureFloodCount :
   void communicateAndMerge();
 
   /**
+   * Sort and assign ids to features based on their positionn in the container after sorting.
```
typo `positionn`

dschwen commented on modules/phase_field/include/postprocessors/FeatureFloodCount.h:
```diff
@@ -306,6 +328,18 @@ class FeatureFloodCount :
+  void scatterAndUpdateRanks();
+
+  /**
    * This routine populatess a stacked vector of local to global indices per rank and the associated count
```
Did you enable sticky keys?

permcody committed:
Address comments

permcody commented:
All comments addressed, tests passed

permcody commented:
Nudge @dschwen...

dschwen commented on modules/phase_field/src/ics/ReconVarIC.C:
```diff
@@ -75,6 +80,21 @@ ReconVarIC::initialSetup()
+    auto index = _consider_phase ? local_id : global_id;
+
+    mooseAssert(index < _assigned_op.size(), "EBSD Global ID is out of bounds");
+    if (index < _assigned_op.size())
```
you have an assert _and_ a runtime check for the same thing here

dschwen commented on modules/phase_field/src/postprocessors/FeatureFloodCount.C:
```diff
@@ -296,43 +298,53 @@ void FeatureFloodCount::communicateAndMerge()
+   * with each feature's _var_index.
+   */
+  unsigned int feature_offset = 0;
+  for (auto map_num = decltype(_maps_size)(0); map_num < _maps_size; ++map_num)
```
beginIndex?

dschwen replied:
like `_feature_counts_per_map` which you use the counter to index into

dschwen commented on modules/phase_field/src/postprocessors/FeatureFloodCount.C:
```diff
@@ -341,12 +353,14 @@ FeatureFloodCount::buildLocalToGlobalIndices(std::vector<unsigned int> & local_t
   mooseAssert(_is_master, "This method must only be called on the root processor");
 
   counts.resize(_n_procs, 0);
```
resize only initializes the new entries with 0. It is not obvious from the code below if each vector element gets written to. For clarity I'd suggest either removing th `, 0` or changing the `resize` to an `assign`, depending on whether you need the zeroing.

permcody replied:
Good catch, definitely need the zeros

dschwen commented on modules/phase_field/src/postprocessors/FeatureFloodCount.C:
```diff
@@ -341,12 +353,14 @@ FeatureFloodCount::buildLocalToGlobalIndices(std::vector<unsigned int> & local_t
 
+  // Build the offsets vector
   unsigned int globalsize = 0;
   std::vector<int> offsets(_n_procs);
```
why not `unsigned int`?

permcody replied:
The MPI interface where this goes requires signed int here. I can add a comment

dschwen commented on modules/phase_field/src/postprocessors/FeatureFloodCount.C:
```diff
@@ -1191,9 +1269,15 @@ operator<<(std::ostream & out, const FeatureFloodCount::FeatureData & feature)
 {
   static const bool debug = true;
```
`static const`?!

permcody replied:
It's a local member purely for debugging and since it's const it doesn't really matter that it's static.

dschwen commented on modules/phase_field/src/postprocessors/GrainTracker.C:
```diff
@@ -131,6 +164,81 @@ GrainTracker::getThreshold(unsigned int current_idx, bool active_feature) const
+    }
+  }
+  else
+    // Just use normally variable inspection on subsequent steps
```
normally -> normal

dschwen commented on modules/phase_field/src/postprocessors/GrainTracker.C:
```diff
@@ -662,18 +707,49 @@ GrainTracker::remapGrains()
+    // Make sure that all split pieces of an EBSD grain are on the same OP
+    if (_ebsd_reader)
+    {
+      for (auto i = beginIndex(_feature_sets); i < _feature_sets.size(); ++i)
```
below you use a different pattern for the same thing (range base for with pointer comparison instead of index loop with index comparison)

permcody replied:
Almost! In this loop I actually need the indices in line 735. In the second loop I never need the indices.

dschwen commented on modules/phase_field/src/postprocessors/GrainTracker.C:
```diff
@@ -1180,28 +1285,28 @@ GrainTracker::updateFieldInfo()
+        _feature_maps[map_idx][entity] = grain._id;
         if (_var_index_mode)
-          _var_index_maps[map_idx][entity] = grain_pair.second._var_idx;
+            _var_index_maps[map_idx][entity] = grain._var_idx;
```
gratuitous indent?

dschwen commented on modules/phase_field/src/postprocessors/GrainTracker.C:
```diff
@@ -1340,19 +1445,34 @@ GrainTracker::boundingRegionDistance(std::vector<MeshTools::BoundingBox> & bboxe
+   * Get the next unique grain ID but make sure to respect
+   * reserve ids.
+   */
+  while (++_max_curr_grain_id <= _reserve_grain_first_idx + _n_reserve_ops)
```
How is that different from 
```
_max_curr_grain_id = _reserve_grain_first_idx + _n_reserve_ops + 1;
```
?!

dschwen replied:
Except being more convoluted of course...

permcody replied:
Not quite! Your statement is not equivalent because _max_curr_grain_id might be greater than the lhs. However this would work too:
_max_curr_grain_id = std::max(_max_curr_grain+id + 1, _reserve_grain_first_idx + _n_reserve_ops);

dschwen replied:
no: `_max_curr_grain_id = std::max(_max_curr_grain+id + 1, _reserve_grain_first_idx + _n_reserve_ops +1);`!

permcody replied:
Nope - the +1 isn't necessary in the second expression because `_reserve_grain_first_idx` is a valid index.

e.g.
Let's say I have 42 grains and I have 2 reserved ops. My grains will have indices from [0, 41]. The reserve grain indices will be 42 and 43. The first time I assign a move a new grain from the reserve op, it needs to be index number 44. I assign `_reserve_grain_first_idx` in line 485:

```
    _reserve_grain_first_idx = _max_curr_grain_id + 1;
   // 42 = 41 + 1
```
Therefore the first index for a new grain is _reserve_grain_first_idx (42) + _n_reserve_ops (2).

permcody replied:
I'll add a comment to explain what I'm doing there so it doesn't trip anyone else up. It's easier to assign the +1 to the `_reserve_grain_first_idx` variable so I don't have to do it every time I need to report that information.

dschwen replied:
your `while (++_max_curr_grain_id <= _reserve_grain_first_idx + _n_reserve_ops);` keeps going if

`_max_curr_grain_id == _reserve_grain_first_idx + _n_reserve_ops`. It quits once `_max_curr_grain_id == _reserve_grain_first_idx + _n_reserve_ops +1`!

dschwen replied:
I do agree, though, that the +1 should not be there (it is just not what your previous code did).

dschwen commented on modules/phase_field/src/postprocessors/GrainTrackerInterface.C:
```diff
@@ -21,6 +21,8 @@ InputParameters validParams<GrainTrackerInterface>()
   params.addParam<UserObjectName>("ebsd_reader", "Optional: EBSD Reader for initial condition");
+  params.addParam<unsigned int>("phase", "EBSD phase number from which to retrieve information");
+
 
```
rm empty line

permcody committed:
Address new round of comments

dschwen commented:
:+1: looks good to me. I'll merge once this passes.
